src/Drips.sol

```diff
@@ -0,0 +1,501 @@
+    /// This function can be used to detect that there are
+    /// too many cycles to analyze in a single transaction.
+    /// @param userId The user ID
+    /// @param assetId The used asset ID
+    /// @return cycles The number of cycles which can be flushed
+    function receivableDripsCycles(
+        Storage storage s,
+        uint64 cycleSecs,
+        uint256 userId,
+        uint256 assetId
+    ) internal view returns (uint64 cycles) {
+        uint64 collectedCycle = s.dripsStates[userId][assetId].nextCollectedCycle;
+        if (collectedCycle == 0) return 0;
+        uint64 currFinishedCycle = _currTimestamp() / cycleSecs;
+        return currFinishedCycle + 1 - collectedCycle;
+    }
+
+    /// @notice Calculate effects of calling `receiveDrips` with the given parameters.
+    /// @param userId The user ID
+    /// @param assetId The used asset ID
+    /// @param maxCycles The maximum number of received drips cycles.
+    /// If too low, receiving will be cheap, but may not cover many cycles.
+    /// If too high, receiving may become too expensive to fit in a single transaction.
+    /// @return receivableAmt The amount which would be received
+    /// @return receivableCycles The number of cycles which would still be receivable after the call
+    function receivableDrips(
+        Storage storage s,
+        uint64 cycleSecs,
+        uint256 userId,
+        uint256 assetId,
+        uint64 maxCycles
+    ) internal view returns (uint128 receivableAmt, uint64 receivableCycles) {
+        uint64 allReceivableCycles = receivableDripsCycles(s, cycleSecs, userId, assetId);
+        uint64 receivedCycles = maxCycles < allReceivableCycles ? maxCycles : allReceivableCycles;
+        receivableCycles = allReceivableCycles - receivedCycles;
+        DripsState storage dripsState = s.dripsStates[userId][assetId];
+        uint64 collectedCycle = dripsState.nextCollectedCycle;
+        int128 cycleAmt = 0;
+        for (uint256 i = 0; i < receivedCycles; i++) {
+            cycleAmt += dripsState.amtDeltas[collectedCycle].thisCycle;
+            receivableAmt += uint128(cycleAmt);
+            cycleAmt += dripsState.amtDeltas[collectedCycle].nextCycle;
+            collectedCycle++;
+        }
+    }
+
+    /// @notice Receive drips from uncollected cycles of the user.
+    /// Received drips cycles won't need to be analyzed ever again.
+    /// Calling this function does not collect but makes the funds ready to be split and collected.
+    /// @param userId The user ID
+    /// @param assetId The used asset ID
+    /// @param maxCycles The maximum number of received drips cycles.
+    /// If too low, receiving will be cheap, but may not cover many cycles.
+    /// If too high, receiving may become too expensive to fit in a single transaction.
+    /// @return receivedAmt The received amount
+    /// @return receivableCycles The number of cycles which still can be received
+    function receiveDrips(
+        Storage storage s,
```

Comment:
Yes, that would be too much logic in `dripsHub`, it would make the library barely reusable in other projects, it would require a lot of expertise but blindly passing a black box `storage` pointer is easy to do. I wouldn't call it a hack, passing a pointer is THE way of accessing storage from libraries, that's in the language design.